spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/MappedToGettableDataConverter.scala

```diff
@@ -0,0 +1,187 @@
+import com.datastax.spark.connector.cql.StructDef
+import com.datastax.spark.connector.mapper.{DefaultColumnMapper, TupleColumnMapper, JavaBeanColumnMapper, ColumnMapper}
+import com.datastax.spark.connector.types.{TupleType, MapType, SetType, ListType, ColumnType, TypeConverter}
+
+private[connector] object MappedToGettableDataConverter {
+
+  def apply[T : TypeTag : ColumnMapper]
+      (struct: StructDef, columnSelection: IndexedSeq[ColumnRef]): TypeConverter[struct.ValueRepr] =
+
+    new TypeConverter[struct.ValueRepr] {
+
+      /** To determine which default column mapper to use for UDT fields.
+        * If this class is a Java bean, then the UDT fields will be mapped with
+        * the `JavaBeanColumnMapper` */
+      private val isJavaBean =
+        implicitly[ColumnMapper[T]].isInstanceOf[JavaBeanColumnMapper[_]]
+
+      private val columnMap =
+        implicitly[ColumnMapper[T]].columnMapForWriting(struct, columnSelection)
+
+      /** Returns the column mapper associated with the given type.
+        * Used to find column mappers for UDT columns.
+        * For tuples, returns [[com.datastax.spark.connector.mapper.TupleColumnMapper TupleColumnMapper]],
+        * for Java beans uses
+        * [[com.datastax.spark.connector.mapper.JavaBeanColumnMapper JavaBeanColumnMapper]],
+        * and for everything else uses
+        * [[com.datastax.spark.connector.mapper.DefaultColumnMapper DefaultColumnMapper]] */
+      private def columnMapper[U: TypeTag]: ColumnMapper[U] = {
+        val tpe = TypeTag.synchronized(typeTag[U].tpe)
+        if (ReflectionUtil.isScalaTuple(tpe))
+          new TupleColumnMapper[U]
+        else if (isJavaBean)
+          new JavaBeanColumnMapper[U]()(ReflectionUtil.classTag[U])
+        else
+          new DefaultColumnMapper[U]
+      }
+
+      /** For these types we don't want to use column mappers */
+      private val dynamicStructRepresentations = Set(
+        TupleValue.Symbol.fullName,
+        UDTValue.Symbol.fullName,
+        "com.datastax.spark.connector.japi.TupleValue",
+        "com.datastax.spark.connector.japi.UDTValue")
+
+      /** Returns true for structs that can store dynamic set of fields which we don't
+        * know Scala types at the moment of searching for the converters. We must not
+        * attempt to obtain ColumnMappers for these types. */
+      private def isDynamicStructRepresentation(sym: Symbol): Boolean =
+        dynamicStructRepresentations.contains(sym.fullName)
+
+      /** Returns true for tuple types that provide full type information of their components */
+      private def isTypedTuple(sym: Symbol): Boolean =
+        ReflectionUtil.isScalaTuple(sym) ||
+          sym == Symbols.PairSymbol ||
+          sym == Symbols.TripleSymbol
+
+      /** Returns a converter for converting given column to appropriate type savable to Cassandra */
+      private def converter[U : TypeTag](columnType: ColumnType[_]): TypeConverter[_ <: AnyRef] = {
```

Comment:
Yeah, big, but fortunately quite flat. This really belongs to the TypeConverter module, but this would need a bigger refactoring. Can we unify this later?